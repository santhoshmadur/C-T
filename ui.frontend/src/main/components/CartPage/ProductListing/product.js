/*******************************************************************************
 *
 *    Copyright 2021 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { gql } from '@apollo/client';
import { useProduct } from '@magento/peregrine/lib/talons/CartPage/ProductListing/useProduct';
import Price from '@magento/venia-ui/lib/components/Price';

import { useStyle } from '@magento/venia-ui/lib/classify';
import Image from '@magento/venia-ui/lib/components/Image';
import Kebab from '@magento/venia-ui/lib/components/LegacyMiniCart/kebab';
import ProductOptions from '@magento/venia-ui/lib/components/LegacyMiniCart/productOptions';
import Section from '@magento/venia-ui/lib/components/LegacyMiniCart/section';
import Quantity from '@magento/venia-ui/lib/components/CartPage/ProductListing/quantity';

import defaultClasses from '@magento/venia-ui/lib/components/CartPage/ProductListing/product.css';

import { CartPageFragment } from '@magento/venia-ui/lib/components/CartPage/cartPageFragments.gql';
import { AvailableShippingMethodsCartFragment } from '@magento/venia-ui/lib/components/CartPage/PriceAdjustments/ShippingMethods/shippingMethodsFragments.gql';

const IMAGE_SIZE = 100;

const Product = props => {
    const { item } = props;

    const { formatMessage } = useIntl();
    const talonProps = useProduct({
        operations: {
            removeItemMutation: REMOVE_ITEM_MUTATION,
            updateItemQuantityMutation: UPDATE_QUANTITY_MUTATION
        },
        ...props,
        item: {
            ...item,
            id: parseInt(item.id)
        }
    });

    const {
        errorMessage,
        handleEditItem,
        handleRemoveFromCart,
        handleUpdateItemQuantity,
        isEditable,
        product,
        isProductUpdating
    } = talonProps;

    const { currency, image, name, options, quantity, stockStatus, unitPrice } = product;

    const classes = useStyle(defaultClasses, props.classes);

    const itemClassName = isProductUpdating ? classes.item_disabled : classes.item;

    const editItemSection = isEditable ? (
        <Section
            text={formatMessage({
                id: 'product.editItem',
                defaultMessage: 'Edit item'
            })}
            onClick={handleEditItem}
            icon="Edit2"
            classes={{
                text: classes.sectionText
            }}
        />
    ) : null;

    const stockStatusMessage =
        stockStatus === 'OUT_OF_STOCK'
            ? formatMessage({
                  id: 'product.outOfStock',
                  defaultMessage: 'Out-of-stock'
              })
            : '';

    return (
        <li className={classes.root}>
            <span className={classes.errorText}>{errorMessage}</span>
            <div className={itemClassName}>
                <div className={classes.imageContainer}>
                    <Image
                        alt={name}
                        classes={{
                            root: classes.imageRoot,
                            image: classes.image
                        }}
                        width={IMAGE_SIZE}
                        resource={image}
                    />
                </div>
                <div className={classes.details}>
                    <div className={classes.name}>{name}</div>
                    <ProductOptions
                        options={options}
                        classes={{
                            options: classes.options,
                            optionLabel: classes.optionLabel
                        }}
                    />
                    <span className={classes.price}>
                        <Price currencyCode={currency} value={unitPrice} />
                        <FormattedMessage id={'product.price'} defaultMessage={' ea.'} />
                    </span>
                    <span className={classes.stockStatusMessage}>{stockStatusMessage}</span>
                    <div className={classes.quantity}>
                        <Quantity itemId={item.id} initialValue={quantity} onChange={handleUpdateItemQuantity} />
                    </div>
                </div>
                <Kebab
                    classes={{
                        root: classes.kebab
                    }}
                    disabled={true}
                >
                    {editItemSection}
                    <Section
                        text={formatMessage({
                            id: 'product.removeFromCart',
                            defaultMessage: 'Remove from cart'
                        })}
                        onClick={handleRemoveFromCart}
                        icon="Trash"
                        classes={{
                            text: classes.sectionText
                        }}
                    />
                </Kebab>
            </div>
        </li>
    );
};

export default Product;

export const REMOVE_ITEM_MUTATION = gql`
    mutation removeItem($cartId: String!, $itemId: Int!) {
        removeItemFromCart(input: { cart_id: $cartId, cart_item_id: $itemId }) @connection(key: "removeItemFromCart") {
            cart {
                id
                ...CartPageFragment
                ...AvailableShippingMethodsCartFragment
            }
        }
    }
    ${CartPageFragment}
    ${AvailableShippingMethodsCartFragment}
`;

export const UPDATE_QUANTITY_MUTATION = gql`
    mutation updateItemQuantity($cartId: String!, $itemId: Int!, $quantity: Float!) {
        updateCartItems(input: { cart_id: $cartId, cart_items: [{ cart_item_id: $itemId, quantity: $quantity }] })
            @connection(key: "updateCartItems") {
            cart {
                id
                ...CartPageFragment
                ...AvailableShippingMethodsCartFragment
            }
        }
    }
    ${CartPageFragment}
    ${AvailableShippingMethodsCartFragment}
`;
