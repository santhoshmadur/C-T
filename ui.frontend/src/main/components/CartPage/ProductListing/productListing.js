/*******************************************************************************
 *
 *    Copyright 2021 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React, { Fragment, Suspense } from 'react';
import { FormattedMessage } from 'react-intl';
import { gql } from '@apollo/client';
import { useProductListing } from '@magento/peregrine/lib/talons/CartPage/ProductListing/useProductListing';

import { useStyle } from '@magento/venia-ui/lib/classify';
import LoadingIndicator from '@magento/venia-ui/lib/components/LoadingIndicator';
import defaultClasses from '@magento/venia-ui/lib/components/CartPage/ProductListing/productListing.css';
import Product from './product';
import { ProductListingFragment } from '@magento/venia-ui/lib/components/CartPage/ProductListing/productListingFragments';

const EditModal = React.lazy(
    async () => await import('@magento/venia-ui/lib/components/CartPage/ProductListing/EditModal').default
);
/**
 * A child component of the CartPage component.
 * This component renders the product listing on the cart page.
 *
 * @param {Object} props
 * @param {Function} props.setIsCartUpdating Function for setting the updating state of the cart.
 * @param {Object} props.classes CSS className overrides.
 * See [productListing.css]{@link https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/CartPage/ProductListing/productListing.css}
 * for a list of classes you can override.
 *
 * @returns {React.Element}
 *
 * @example <caption>Importing into your project</caption>
 * import ProductListing from "@magento/venia-ui/lib/components/CartPage/ProductListing";
 */
const ProductListing = props => {
    const { onAddToWishlistSuccess, setIsCartUpdating, fetchCartDetails } = props;
    const talonProps = useProductListing({
        queries: {
            getProductListing: GET_PRODUCT_LISTING
        }
    });
    const { activeEditItem, isLoading, items, setActiveEditItem, wishlistConfig } = talonProps;

    const classes = useStyle(defaultClasses, props.classes);

    if (isLoading) {
        return (
            <LoadingIndicator>
                <FormattedMessage id={'productListing.loading'} defaultMessage={'Fetching Cart...'} />
            </LoadingIndicator>
        );
    }

    if (items.length) {
        const productComponents = items.map(product => (
            <Product
                item={product}
                key={product.id}
                setActiveEditItem={setActiveEditItem}
                setIsCartUpdating={setIsCartUpdating}
                onAddToWishlistSuccess={onAddToWishlistSuccess}
                fetchCartDetails={fetchCartDetails}
                wishlistConfig={wishlistConfig}
            />
        ));

        return (
            <Fragment>
                <ul className={classes.root}>{productComponents}</ul>
                <Suspense fallback={null}>
                    <EditModal
                        item={activeEditItem}
                        setIsCartUpdating={setIsCartUpdating}
                        setActiveEditItem={setActiveEditItem}
                    />
                </Suspense>
            </Fragment>
        );
    }

    return null;
};

export const GET_PRODUCT_LISTING = gql`
    query getProductListing($cartId: String!) {
        cart(cart_id: $cartId) {
            id
            ...ProductListingFragment
        }
    }
    ${ProductListingFragment}
`;

export default ProductListing;
