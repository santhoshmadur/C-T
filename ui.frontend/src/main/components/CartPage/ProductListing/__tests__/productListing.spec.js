/*******************************************************************************
 *
 *    Copyright 2021 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import LoadingIndicator from '@magento/venia-ui/lib/components/LoadingIndicator';
import { useProductListing } from '@magento/peregrine/lib/talons/CartPage/ProductListing/useProductListing';

import ProductListing from '../productListing';
import render from '../../../utils/test-utils';

jest.mock('@magento/peregrine/lib/talons/CartPage/ProductListing/useProductListing');
jest.mock('@apollo/client', () => {
    return {
        gql: jest.fn(),
        useLazyQuery: jest.fn()
    };
});

jest.mock('@magento/peregrine/lib/context/cart', () => {
    const state = { cartId: 'cart123' };
    const api = {};
    const useCartContext = jest.fn(() => [state, api]);

    return { useCartContext };
});
jest.mock('react-router-dom', () => ({
    Link: ({ children, ...rest }) => <div {...rest}>{children}</div> // eslint-disable-line react/display-name
}));
jest.mock('@magento/peregrine/lib/util/makeUrl');
jest.mock('../product', () => 'Product');
jest.mock('@magento/venia-ui/lib/components/CartPage/ProductListing/EditModal', () => 'EditModal');

describe('productListing', () => {
    test('renders null with no items in cart', () => {
        useProductListing.mockReturnValueOnce({
            isLoading: false,
            isUpdating: false,
            items: []
        });

        const tree = render(<ProductListing />);

        expect(tree.toJSON()).toMatchSnapshot();
    });

    test('renders list of products with items in cart', () => {
        useProductListing.mockReturnValueOnce({
            isLoading: false,
            isUpdating: false,
            items: ['1', '2', '3']
        });

        const tree = render(<ProductListing />);

        expect(tree.toJSON()).toMatchSnapshot();
    });

    test('renders loading indicator if isLoading', () => {
        useProductListing.mockReturnValueOnce({
            isLoading: true
        });

        const propsWithClass = {
            classes: {
                root: 'root'
            }
        };

        const tree = render(<ProductListing {...propsWithClass} />);

        expect(tree.root.findByType(LoadingIndicator)).toBeTruthy();
    });
});
