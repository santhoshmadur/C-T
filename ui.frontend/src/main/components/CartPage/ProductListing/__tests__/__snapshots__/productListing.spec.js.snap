// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`productListing renders list of products with items in cart 1`] = `
<ul>
  <Product
    item="1"
  />
  <Product
    item="2"
  />
  <Product
    item="3"
  />
</ul>
`;

exports[`productListing renders null with no items in cart 1`] = `null`;
